pic-sure-auth-services/src/main/java/edu/harvard/hms/dbmi/avillach/auth/service/impl/UserService.java

```diff
@@ -0,0 +1,511 @@
+            }
+
+            if (user.getEmail() == null) {
+                try {
+                    logger.info("Parsing metadata for email address");
+                    HashMap<String, String> metadata = new HashMap<String, String>(new ObjectMapper().readValue(user.getGeneralMetadata(), Map.class));
+                    List<String> emailKeys = metadata.keySet().stream().filter((key) -> key.toLowerCase().contains("email")).toList();
+                    if (!emailKeys.isEmpty()) {
+                        user.setEmail(metadata.get(emailKeys.getFirst()));
+                    }
+                } catch (IOException e) {
+                    logger.error("Failed to parse metadata for email address", e);
+                }
+            }
+        }
+
+        users = addUser(users);
+        return users;
+    }
+
+    /**
+     * check all referenced field if they are already in database. If
+     * they are in database, then retrieve it by id, and attach it to
+     * user object.
+     *
+     * @param users A list of users
+     */
+    private void checkAssociation(List<User> users) {
+        for (User user : users) {
+            if (user.getRoles() != null) {
+                Set<Role> roles = new HashSet<>();
+                user.getRoles().forEach(t -> {
+                    Optional<Role> role = this.roleService.getRoleById(t.getUuid());
+                    if (role.isEmpty()) {
+                        throw new RuntimeException("Role not found - uuid: " + t.getUuid().toString());
+                    }
+                    roles.add(role.get());
+                });
+
+                user.setRoles(roles);
+            }
+
+            if (user.getConnection() != null) {
+                Optional<Connection> connection = this.connectionRepository.findById(user.getConnection().getId());
+                user.setConnection(connection.orElse(null));
+            }
+        }
+    }
+
+    @Transactional
+    public List<User> updateUser(List<User> users) {
+        SecurityContext securityContext = SecurityContextHolder.getContext();
+        logger.info("Security context: {}", securityContext.getAuthentication().getPrincipal());
+        CustomUserDetails customUserDetails = (CustomUserDetails) securityContext.getAuthentication().getPrincipal();
+        if (customUserDetails == null || customUserDetails.getUser() == null && customUserDetails.getUser().getUuid() == null) {
+            logger.error("Security context didn't have a user stored.");
+            return null;
+        }
+
+        User currentUser = customUserDetails.getUser();
+        checkAssociation(users);
+        boolean allowUpdate = true;
+        for (User user : users) {
+            Optional<User> originalUser = this.userRepository.findById(user.getUuid());
+            if (!allowUpdateSuperAdminRole(currentUser, user, originalUser.orElse(null))) {
+                allowUpdate = false;
+                break;
+            }
+        }
```

Comment:
Aren't these looping queries a huge smell? I realize `userRepository.findAllById` requires more work to make correct, but this seems expensive.

Reply:
I will do this as well.